src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,18 +31,25 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
+            if (provider == null)
+            {
+                throw new ArgumentNullException(nameof(provider));
+            }
+
             int bestLength = -1;
+            bool multipleBestLengthFound = false;
             bool seenPreferred = false;
 
             ConstructorMatcher bestMatcher = default;
+            IServiceProviderIsService? serviceProviderIsService = provider.GetService<IServiceProviderIsService>();
```

Comment:
We should be careful about relying on an IServiceProviderIsService implementation giving false negatives. This should never happen with the built-in ServiceProvider implementation, but it could happen with third part containers. This turned out to be a major point of contention when [approving the IServiceProviderIsService API](https://github.com/dotnet/runtime/issues/53919#issuecomment-858861117). False positives should not happen with any container.

[Here's a timestamped video of that discussion.](https://www.youtube.com/watch?v=YRcGjsFbx4Y&t=0h55m13s)

I think there should at least be a way for developers to force ActivatorUtilities to use a specific constructor even if an IServiceProviderIsService exists and `IServiceProviderIsService.IsService(parameterType)` returns false. In ASP.NET Core for example, if you use the `[FromServices]` attribute, we will try to resolve the service regardless of what the IServiceProviderIsService has to say about it.

Reply:
@tillig Do you have any thoughts on this?

Reply:
I forgot ActivatorUtilitiesConstructorAttribute exists because we so rarely use it, but maybe that sufficient for this case. And this only affects ties, right? I'll mark this as resolved.

Reply:
Agreed, `IsService` is a "best effort" API, not a guarantee. In Autofac, for example, registrations may come from a dynamic registration source rather than be added explicitly to the container, in which case [at least the Autofac implementation for `IsService` might return `false`](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/bf61f9225b078359c7e9e688b3823990d056a4f2/src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs#L53) even if the target could technically be resolved. It will _never_ return `true` if it can't be resolved.

Reply:
IIRC there was also some discussion about the `IsService` being something that indicates "it's registered" but _not_ that it can necessarily be successfully resolved. All that said, I just did a bunch of testing for the 90% case just now to see what we detect as registered vs. not and we're _usually_ correct and do detect most of the dynamic stuff, too. I know @alistairjevans did a lot of work to help with assembly trimming and may have impacted our accuracy in a positive way here. Regardless, I think you're right that it's likely not a big deal since it is such an edge case.

Reply:
Dynamic sources generally should be ok
..the false negative problem came from the fact that, in Autofac (and a bunch of other IoC containers) you can register service implementations in a nested unit-of-work/request scope, that didn't exist in the root scope.

So, if `IsService` is called from the root scope, and returns false, it could conceivably return true if called from a more nested scope.

For perfect accuracy of checking whether a given method argument is a service, the `IsService` call would need to happen on the same scope the services are about to be resolved from.

Reply:
> For perfect accuracy of checking whether a given method argument is a service, the IsService call would need to happen on the same scope the services are about to be resolved from.

In this case it is. The check for `IsService` happens when trying to select the constructor. Then once a ctor is selected, the same `IServiceProvider` is used to get the services to populate the ctor's arguments.